### PR TITLE
[Feature]Modification of the finetune mechanism

### PR DESCRIPTION
**Pull request by DavdGao**
This PR is mainly for the modification of the finetune mechansim (#148 ), but we also make small change for other functions as following
## Finetune 
- Move partial parameters from `cfg.federate` into `cfg.train` as they are more relevant to the training, including
    - `local_update_step`
    - `batch_or_epoch`
- Creat `cfg.finetune` and `cfg.train` in the config to support different parameters for finetuning and training (e.g. `optimizer.lr`)
- Implement `finetune` function in the basic trainer
- Modify most existing shells and yaml files to fit the new setting (except the files under the directory `benchmark`)
## Enums and Decorators
- Create `enums.py` to avoid using string and the inconsistency issues
- Create `decorators.py` to keep the code clean
## Optimizer
- Initialize the `ctx.optimizer` in the beginning of the routine function rather than within the `context` to solve #136 

## To be discussed
@joneswong please check if the following modifications are appropriate
- In this PR, `use_diff` is implemented by a decorator [`use_diff`](https://github.com/alibaba/FederatedScope/pull/177/files#diff-3aa904a9c4ca6033f418b8d93b000f9102cb6cae2b0dd9085bb7e9b4cbdcd6e6R20).
- Some hpo configs are modifed to fit the new configuration.

### CODE REVIEW

File: federatedscope/core/auxiliaries/decorators.py [hunk 1]
```diff
@@ -0,0 +1,20 @@
+def use_diff(func):
```
[the remark this answers is not in the document]
Comment (rayrayraykk) [reply federatedscope/core/auxiliaries/decorators.py:1]: This part looks good to me.

File: federatedscope/core/trainers/trainer_pFedMe.py [hunk 2]
```diff
@@ -73,16 +73,16 @@ def init_pFedMe_ctx(base_trainer):
...
     # the optimizer used in pFedMe is based on Moreau Envelopes regularization
     # besides, there are two distinct lr for the approximate model and base model
     ctx.optimizer = wrap_regularized_optimizer(
-        ctx.optimizer, cfg.personalization.regular_weight)
+        ctx.optimizer, ctx.cfg.personalization.regular_weight)
     for g in ctx.optimizer.param_groups:
-        g['lr'] = cfg.personalization.lr
-    ctx.pFedMe_outer_lr = cfg.optimizer.lr
+        g['lr'] = ctx.cfg.personalization.lr
+    ctx.pFedMe_outer_lr = ctx.cfg.train.optimizer.lr
```
Comment (xieyxclack) [inline federatedscope/core/trainers/trainer_pFedMe.py:84]: Shall we have `ctx.cfg.xxx` ? @yxdyc

**Review by xieyxclack — COMMENTED** [review]
Good job! It might help us to split the functionalities of the trainer. And @yxdyc can double-check the details of the trainer.
Please note that the modification of config (such as `cfg.optimizer` -> `cfg.traine.optimizer`) would affect a large range of existing files, and we should make sure the provided examples and the declarations have been updated.

File: federatedscope/core/trainers/trainer_pFedMe.py (hunk 2 above)
Comment (yxdyc) [reply federatedscope/core/trainers/trainer_pFedMe.py:84]: Both are ok, as we don't usually change `cfg.personalization.lr` during the FL courses

**Comment by yxdyc** [comment]
LGTM. The modification of configs can invalid the scripts of our three FL benchmarks. We can discuss this later that whether we need to make our three benchmarks compatible with FS 0.2.0.

**Comment by rayrayraykk** [comment]
LGTM, but please rebase the master to request a new CI-test for format checks, thanks.

**Comment by DavdGao** [comment]
> LGTM, but please rebase the master to request a new CI-test for format checks, thanks.

modified accordingly

**Review by xieyxclack — APPROVED** [review]
LGTM

**Pull request merged**